node/full_node_integration_test.go

```diff
@@ -353,6 +353,130 @@ func testSingleAggreatorSingleFullNodeSingleLightNode(t *testing.T) {
 	assert.Equal(n1h, n3h, "heights must match")
 }
 
+func checkAppCalls(assert *assert.Assertions, app *mocks.Application, expected []int) {
+	generateFraudProofCnt := 0
+	verifyFraudProofCnt := 0
+	for _, call := range app.Calls {
+		switch call.Method {
+		case "GenerateFraudProof":
+			generateFraudProofCnt++
+		case "VerifyFraudProof":
+			verifyFraudProofCnt++
+		}
+	}
+	assert.Equal(expected[0], generateFraudProofCnt)
+	assert.Equal(expected[1], verifyFraudProofCnt)
+}
+
+func testSingleAggreatorSingleFullNodeFraudProofGossip(t *testing.T) {
+	assert := assert.New(t)
+	require := require.New(t)
+
+	var wg sync.WaitGroup
+	aggCtx, aggCancel := context.WithCancel(context.Background())
+	ctx, cancel := context.WithCancel(context.Background())
+	clientNodes := 1
+	nodes, apps := createNodes(aggCtx, ctx, clientNodes+1, true, &wg, t)
+
+	for _, app := range apps {
+		app.On("VerifyFraudProof", mock.Anything).Return(abci.ResponseVerifyFraudProof{Success: true}).Run(func(args mock.Arguments) {
+			wg.Done()
+		}).Once()
+	}
+
+	aggNode := nodes[0]
+	fullNode := nodes[1]
+
+	wg.Add(clientNodes + 1)
+	require.NoError(aggNode.Start())
+	time.Sleep(2 * time.Second)
+	require.NoError(fullNode.Start())
+
+	wg.Wait()
+	// aggregator should have 0 GenerateFraudProof calls and 1 VerifyFraudProof calls
+	checkAppCalls(assert, apps[0], []int{0, 1})
+	// fullnode should have 1 GenerateFraudProof calls and 1 VerifyFraudProof calls
+	checkAppCalls(assert, apps[1], []int{1, 1})
+
+	n1Frauds, err := aggNode.fraudService.Get(aggCtx, types.StateFraudProofType)
+	require.NoError(err)
+	aggCancel()
+	require.NoError(aggNode.Stop())
+
+	n2Frauds, err := fullNode.fraudService.Get(aggCtx, types.StateFraudProofType)
+	require.NoError(err)
+	cancel()
+	require.NoError(fullNode.Stop())
+
+	assert.Equal(len(n1Frauds), 1, "number of fraud proofs received via gossip should be 1")
+	assert.Equal(len(n2Frauds), 1, "number of fraud proofs received via gossip should be 1")
+	assert.Equal(n1Frauds, n2Frauds, "the received fraud proofs after gossip must match")
+}
+
+func testSingleAggreatorTwoFullNodeFraudProofSync(t *testing.T) {
+	assert := assert.New(t)
+	require := require.New(t)
+
+	var wg sync.WaitGroup
+	aggCtx, aggCancel := context.WithCancel(context.Background())
+	ctx, cancel := context.WithCancel(context.Background())
+	clientNodes := 2
+	nodes, apps := createNodes(aggCtx, ctx, clientNodes+1, true, &wg, t)
+
+	for _, app := range apps {
+		app.On("VerifyFraudProof", mock.Anything).Return(abci.ResponseVerifyFraudProof{Success: true}).Run(func(args mock.Arguments) {
+			wg.Done()
+		}).Once()
+	}
+
+	aggNode := nodes[0]
+	fullNode1 := nodes[1]
+	fullNode2 := nodes[2]
+
+	wg.Add(clientNodes)
```

Comment:
can we make a follow up issue for why we need to use the wait group like this?
this seems like an unnecessary complexity using a waitgroup like this.

I think a better design would be to just have the start node function return a done chan, then there is no surface area of bugs due to devs not adding the right number `wg.Add`

in general I think waitgroups should only be used when you can see the entire scope, otherwise chans are a better tool.

Reply:
here the purpose of waitgroup is to prevent data race between `checkAppCalls` in the test and `node` which both access `app *mocks.Application`. we want to wait until the node completes update to `app *mocks.Application` and then we call `checkAppCalls`. 

previously we had `chan`, however synchronizing would require multiple channels (between nodes) and was even more complicated to manage race between them. 

i think waitgroup is more appropriate here as we just want to synchronize between two sets 1) nodes and 2) test.

Reply:
if the problem is a data race, why wouldn't we use a mutex to control the impacted field instead of trying to control the timing of the calls?  I'm mostly just curious to understand the flow of the code more as this seems like a weird way to handle this. 

happy to take this offline as it isn't blocking this PR.

Reply:
read the code some more. 

So if I'm understanding correctly, the datarace occurs because `checkAppCalls` is access the `Calls` field of the `app *mocks.Application` without acquiring the mutex. 

This is currently being solved with a waitgroup that forces the test to wait until `we believe` the `Call` field was updated. 

If this is accurate, then I'm doubling down on my original comment about needing a different solution 😄 

Looks like there is already a built in for this `AssertNumberOfCalls` in the mock package. Can't we just use that?

So we should be able to update this test to using the `AssertNumberOfCalls` instead of `checkAppCalls`. 
And if we do it in a retry then you shouldn't need the waitgroups at all and we can remove that whole complexity. 

Please let me know if I'm missing something.

Reply:
@MSevey we can definitely replace custom `checkAppCalls` function with existing `AssertNumberOfCalls`. with respect to assert with retry, do you have some suggestions on what is the standard way to do this?

Reply:
as discussed offline, let's remove `checkAppCalls` in this PR and then create a follow up issue for handling the full removal of the waitgroup and the use of retries.

Reply:
linking newly created issue here: https://github.com/rollkit/rollkit/issues/956